README.md
# 8 Queens

A Delphi Pascal Windows Program that calculates and displays all possible solutions of the 8 Queens Problem

> ***This project is archived and is no longer being developed. No support is available. Bug reports, feature requests and pull requests are not being accepted.***

For more information about the program please see the [8 Queens page](https://delphidabbler.com/software/queens) on [delphidabbler.com](https://delphidabbler.com).

A user guide is available in open document and PDF formats in this repo's `Docs` directory. The user guide is also included in the final release.

## Final release

The final release of the program is [v4.1.1](https://github.com/ddab-archive/8queens/releases/tag/v4.1.1-FINAL).

## Source Code

The source code was not originally developed under version control. Eventually the source code was moved to a Subversion repository. A summary of changes made before moving to version control can be found in [`Docs/PreSVNHistory.txt`](https://raw.githubusercontent.com/ddab-archive/8queens/main/Docs/PreSVNHistory.txt). The remainder of the history was recorded in the Subversion repo, a dump of which has been uploaded to `SVN/8queens-svndump.zip` [[download](https://github.com/ddab-archive/8queens/raw/main/SVN/8queens-svndump.zip)].

### Commits since the last release

The `main` branch of this project differs from the final release in two ways.

1. The original `README.md` file has been renamed `README-original.md` so that this file will be displayed on GitHub.
2. `8queens-svndump.zip` was uploaded to the `SVN` directory, as noted above.

The original `README.md` is the one that is included with the final release. It is also included in the source code zip file for that release.

The source code as it was at the final release can be viewed on GitHub by switching to the [`v4.1.1-FINAL`](https://github.com/ddab-archive/8queens/tree/v4.1.1-FINAL) tag.

## License

This program is released under the terms of the [MIT License](http://delphidabbler.mit-license.org/1991-2016/).


Docs/ChangeLog.txt
; ------------------------------------------------------------------------------
; Change Log for 8 Queens
;
; $Rev: 82 $
; $Date: 2016-02-15 00:55:02 +0000 (Mon, 15 Feb 2016) $
; ------------------------------------------------------------------------------

Release v4.1.1 of 15 February 2016
+ Fixed memory leak. Achieved by recompiling with bug-fixed v2.2.1 of the DelphiDabbler Hot Label Component, an earlier version of which caused the leak.
+ Added Windows compatibility information to application manifest indicating compatibility with Windows 10 and lower.
+ Now compiled with Delphi XE instead of Delphi 2010. 
+ Thoroughly overhauled and reorganised documentation, including an update to the user manual.
+ Re-licensed program under MIT license: http://delphidabbler.mit-license.org/1991-2016/.
+ Installer no longer gives an option to display README file.

Release v4.1 of 18 March 2012
+ Holding mouse key down on a square now displays rows, columns and diagonals in which a queen on the square can take other pieces.
+ Some minor UI changes, slightly increasing size of chessboard, prompt area and main window.
+ Significant refactoring, particularly to problem solving engine and chessboard display.
+ Modified to compile with Delphi 2010. Removed display bug fix code added in v4.0.2 since not required with Delphi 2010.
+ Installer changed to be Unicode compatible.
+ Updated documentation, including user guide and on-screen prompts.
+ Most source code changed to Mozilla Public License v2.0.

Release v4.0.2 of 18 June 2008
+ Provided work-around for Delphi's Alt key bug on XP and Vista (CodeGear Quality Central bug report #374030).
+ Changed to make application minimisation, task bar preview window, and appearance in "Flip 3D" task switching display correctly on Windows Vista.
+ Added support for default operating system font and adjusted controls for larger Vista Segoe font.
+ Updated manifest resource to ensure compatibility with Windows Vista. Continued to use themes.
+ Some minor refactoring of calculation engine.
+ Revised install script to use macros.

Release v4.0.1 of 05 March 2007
+ Revised main form so that only chess board itself is redrawn when we need to update it, rather than whole form.
+ Corrected and rephrased some prompt text.
+ Added new batch file to build program.
+ Created new installer using Inno Setup.
+ Changed to new end user license agreement for the executable program. The program remains open source.

Release v4.0 of 06 November 2005
+ Converted to 32 bit Windows with XP theme support.
+ Redesigned main window.
+ Redesigned solution engine as state machine with facility to reset the engine to enable all solutions to be found again.
+ Added facility to restart finding of solutions.
+ Added ability to automatically show solutions as a "slide show".
+ Removed Windows help file and replaced with context sensitive prompts displayed in main window.
+ Added link to DelphiDabbler.com.
+ Changed to Mozilla Public License.

v3.0 of 03 May 1998, not released until 01 April 1999
+ Major re-write from v2.0.
+ Main user interface redesigned to appear as a dialog box.
+ The queens on the chessboard now appear as images of chess pieces rather than blobs.
+ Help system added which tells how to use the program and also how the problem was solved. Help file also includes the full source code.
+ About box added which reads its information from a version information resource included in the program file.
+ Problem solution engine re-written in an object oriented way.

Un-released v2.1 of 21 July 1997
+ Added application title to project file.
+ Added custom program icon.
+ Removed design-time caption from form and set caption at run-time to be same as application title.

Un-released v2.0 of 06 March 1997
+ Total rewrite as Windows 3.1 program - compiled with Delphi 1.

v1.0 of 09 December 1991, not released until 01 April 1999
+ Original DOS version - compiled with Turbo Pascal.


Source/UPlatform.pas
{
 * Distributed under the MIT license.
 * See the accompanying LICENSE file or go to
 * http://delphidabbler.mit-license.org/1991-2016/
 *
 * $Rev: 66 $
 * $Date: 2016-02-14 02:15:25 +0000 (Sun, 14 Feb 2016) $
 *
 * Provides OS platform specific information and customisations.
}


unit UPlatform;


interface


uses
  // Delphi
  Forms;


///  <summary>Sets default font of given form and its contained controls to use
///  OSs default system font if necessary.</summary>
procedure SetDefaultFormFont(const AForm: TCustomForm);


implementation


uses
  // Delphi
  SysUtils, Classes, Controls, Windows, Graphics;


type
  ///  <summary>Class helper that provides access to protected Font and
  ///  ParentFont properties of TControl.</summary>
  TControlHelper = class helper for TControl
  public
    ///  <summary>Gets value of TControl.Font property.</summary>
    function GetFont: TFont;
    ///  <summary>Gets value of TControl.ParentFont property.</summary>
    function GetParentFont: Boolean;
  end;

///  <summary>Helper function that gets handle to underlying operating system's
///  default font.</summary>
function GetDefaultFontHandle: HFONT;
var
  LogFont: TLogFont;  // structure storing info about logical font
begin
  if SystemParametersInfo(
    SPI_GETICONTITLELOGFONT, SizeOf(LogFont), @LogFont, 0
  ) then
    // system font is same as that used for desktop icon titles
    Result := CreateFontIndirect(LogFont)
  else
    // can't get icon title font: use stock object
    Result := GetStockObject(DEFAULT_GUI_FONT);
end;

procedure SetDefaultFormFont(const AForm: TCustomForm);
var
  DefFont: TFont;   // operating system's default font
  CmpIdx: Integer;  // loops through all form's owned components
  Cmp: TComponent;  // reference to each component owned by form
  Ctrl: TControl;
begin
  // Create default font
  DefFont := TFont.Create;
  try
    DefFont.Handle := GetDefaultFontHandle;
    // Scan through all form's components, looking for TControls
    for CmpIdx := 0 to Pred(AForm.ComponentCount) do
    begin
      Cmp := AForm.Components[CmpIdx];
      if (Cmp is TControl) then
      begin
        Ctrl := Cmp as TControl;
        // Got a TControl. If control's font is not same as form's, but has same
        // name, we change name to default font name. If control's font has same
        // name and size as form, we change both name and size to that of
        // default font
        if not Ctrl.GetParentFont and
          AnsiSameText(Ctrl.GetFont.Name, AForm.Font.Name) then
        begin
          if Ctrl.GetFont.Size = AForm.Font.Size then
            Ctrl.GetFont.Size := DefFont.Size;
          Ctrl.GetFont.Name := DefFont.Name;
        end;
      end;
    end;
    // Finally, change form's font: this affects all controls with ParentFont
    // equal to True
    AForm.Font.Assign(DefFont);
  finally
    FreeAndNil(DefFont);
  end;
end;

{ TControlHelper }

function TControlHelper.GetFont: TFont;
begin
  Result := Self.Font;
end;

function TControlHelper.GetParentFont: Boolean;
begin
  Result := Self.ParentFont;
end;

end.



Source/UChessBoardCmp.pas
{
 * Distributed under the MIT license.
 * See the accompanying LICENSE file or go to
 * http://delphidabbler.mit-license.org/1991-2016/
 *
 * $Rev: 66 $
 * $Date: 2016-02-14 02:15:25 +0000 (Sun, 14 Feb 2016) $
 *
 * Custom component that displays chess board and its contents.
}


unit UChessBoardCmp;


interface


uses
  // Delphi
  Classes, Controls, Windows, Graphics,
  // Project
  UCalc;

type
  ///  <summary>Range of permitted sizes of chessboard squares.</summary>
  TSquareSize = 20..36;

type
  ///  <summary>Control that displays chessboard, queens and lines of queens
  ///  influence.</summary>
  TChessBoardCmp = class(TCustomControl)
  strict private
    ///  <summary>Records state of solution engine.</summary>
    fEngineState: TEngineState;
    ///  <summary>Bitmap storing image of queen.</summary>
    fQueenBmp: TBitmap;
    ///  <summary>Flag indicating if lines showing influences of square recorded
    ///  in fInfluencesSquare are to be displayed.</summary>
    fShowInfluences: Boolean;
    ///  <summary>Square for which influence lines are to be displayed.
    ///  </summary>
    ///  <remarks>Value is undefined if fShowInfluences is False.</remarks>
    fInfluencesSquare: TSquare;
    ///  <summary>Value of SquareSize property.</summary>
    fSquareSize: TSquareSize;
    ///  <summary>Returns coordinates of chessboard square under mouse
    ///  cursor at client coordinates (X,Y).</summary>
    function SquareCoordsAt(const X, Y: Integer): TPoint;
    ///  <summary>Paints chessboard squares on control's canvas.</summary>
    procedure PaintBoard;
    ///  <summary>Paints queens onto control's canvas.</summary>
    procedure PaintQueens;
    ///  <summary>Paints lines of influence of square stored in
    ///  fInfluencesSquare.</summary>
    ///  <remarks>Must not be called if fShowInfluences is False.</remarks>
    procedure PaintInfluences;
    ///  <summary>Write accessor for SquareSize property.</summary>
    ///  <remarks>Redraws display if value changed.</remarks>
    procedure SetSquareSize(const Value: TSquareSize);
    ///  <summary>Gets size of one side of the chessboard in pixels.</summary>
    function BoardSize: Integer;
  protected
    ///  <summary>Paints control onto its canvas.</summary>
    procedure Paint; override;
  public
    ///  <summary>Constructs and initialises control.</summary>
    constructor Create(AOwner: TComponent); override;
    ///  <summary>Tears down control and tidies up.</summary>
    destructor Destroy; override;
    ///  <summary>Sets bounds of control, ignoring AWidth and AHeight and
    ///  preserving board size.</summary>
    procedure SetBounds(ALeft, ATop, AWidth, AHeight: Integer); override;
    ///  <summary>Updates display to reflect given solution engine state.
    ///  </summary>
    procedure UpdateBoard(const EngineState: TEngineState);
    ///  <summary>Displays influence lines for given square.</summary>
    procedure ShowInfluences(const Square: TSquare);
    ///  <summary>Hides any displayed influence lines.</summary>
    procedure HideInfluences;
    ///  <summary>Returns chessboard square under mouse cursor at client
    ///  coordinates (X,Y).</summary>
    ///  <remarks>Requires that there is a square under the mouse.</remarks>
    function GetSquareAt(const X, Y: Integer): TSquare;
    ///  <summary>Checks if there is a chessboard square under the mouse.
    ///  </summary>
    function IsSquareAt(const X, Y: Integer): Boolean;
  published
    ///  <summary>OnMouseDown event made public.</summary>
    property OnMouseDown;
    ///  <summary>OnMouseUp event made public.</summary>
    property OnMouseUp;
    ///  <summary>Size of a chessboard square in pixels.</summary>
    ///  <remarks>Setting this property changes the size of the control.
    ///  </remarks>
    property SquareSize: TSquareSize read fSquareSize write SetSquareSize
      default 22;
  end;

implementation


uses
  // Delphi
  Types;


{ TChessBoardCmp }

function TChessBoardCmp.BoardSize: Integer;
begin
  Result := (High(TBoardCoord) - Low(TBoardCoord) + 1) * fSquareSize;
end;

constructor TChessBoardCmp.Create(AOwner: TComponent);
begin
  inherited;
  fQueenBmp := TBitmap.Create;
  fQueenBmp.Handle := LoadBitmap(HInstance, 'QUEEN');
  fSquareSize := 22;
  Width := BoardSize;
  Height := BoardSize;
  DoubleBuffered := True;
end;

destructor TChessBoardCmp.Destroy;
begin
  fQueenBmp.Free;
  inherited;
end;

function TChessBoardCmp.GetSquareAt(const X, Y: Integer): TSquare;
var
  Coords: TPoint; // coordinates of square under mouse cursor
begin
  Assert(IsSquareAt(X, Y));
  Coords := SquareCoordsAt(X, Y);
  Result := TSquare.Create(Coords.X, Coords.Y);
end;

procedure TChessBoardCmp.HideInfluences;
begin
  if fShowInfluences then
  begin
    fShowInfluences := False;
    Invalidate;
  end;
end;

function TChessBoardCmp.IsSquareAt(const X, Y: Integer): Boolean;
var
  Coords: TPoint; // coordinates of square under mouse cursor
begin
  Coords := SquareCoordsAt(X, Y);
  Result := (Coords.X in [Low(TColCoord)..High(TColCoord)])
    and (Coords.Y in [Low(TRowCoord)..High(TRowCoord)]);
end;

procedure TChessBoardCmp.Paint;
begin
  inherited;
  PaintBoard;
  PaintQueens;
  if fShowInfluences then
    PaintInfluences;
end;

procedure TChessBoardCmp.PaintBoard;
var
  TopLeft: TPoint;  // coordinates of top left of square being drawn
  Col: TColCoord;   // loops thru each column of chessboard
  Row: TRowCoord;   // loops thru each row of chessboard
begin
  // Draw grid
  Canvas.Brush.Style := bsSolid;
  Canvas.Pen.Style := psClear;
  for Col := Low(TColCoord) to High(TColCoord) do
  begin
    for Row := Low(TRowCoord) to High(TRowCoord) do
    begin
      if Odd(Col) = Odd(Row) then
        Canvas.Brush.Color := clBlack
      else
        Canvas.Brush.Color := clWhite;
      TopLeft := Point(SquareSize * (Col - 1), SquareSize * (Row - 1));
      Canvas.Rectangle(Bounds(TopLeft.X, TopLeft.Y, SquareSize, SquareSize));
    end;
  end;
end;

procedure TChessBoardCmp.PaintInfluences;
var
  Square: TInfluencedSquare;  // each influenced square
  TopLeft: TPoint;            // coordinate of top left corner of each square
begin
  Canvas.Brush.Style := bsClear;
  Canvas.Pen.Color := clRed;
  Canvas.Pen.Style := psDot;
  for Square in fEngineState.InfluencedSquares(
    fInfluencesSquare.Col, fInfluencesSquare.Row
  ) do
  begin
    TopLeft := Point(
      SquareSize * (Square.Square.Col - 1), SquareSize * (Square.Square.Row - 1)
    );
    if siColumn in Square.Influences then
    begin
      Canvas.MoveTo(TopLeft.X + SquareSize div 2, TopLeft.Y);
      Canvas.LineTo(TopLeft.X + SquareSize div 2, TopLeft.Y + SquareSize);
    end;
    if siRow in Square.Influences then
    begin
      Canvas.MoveTo(TopLeft.X, TopLeft.Y + SquareSize div 2);
      Canvas.LineTo(TopLeft.X + SquareSize, TopLeft.Y + SquareSize div 2);
    end;
    if siUpDiag in Square.Influences then
    begin
      Canvas.MoveTo(TopLeft.X + SquareSize, TopLeft.Y);
      Canvas.LineTo(TopLeft.X, TopLeft.Y + SquareSize);
    end;
    if siDownDiag in Square.Influences then
    begin
      Canvas.MoveTo(TopLeft.X + SquareSize, TopLeft.Y + SquareSize);
      Canvas.LineTo(TopLeft.X, TopLeft.Y);
    end;
  end;
end;

procedure TChessBoardCmp.PaintQueens;
var
  QueenOffset: TPoint;  // offset of queen image in a square
  QueenRect: TRect;     // bounding rectangle of queen bitmap
  Square: TSquare;      // each queen square on board
  TopLeft: TPoint;      // top left corner of each queen square
begin
  QueenOffset := Point(
    (SquareSize - fQueenBmp.Width) div 2,
    (SquareSize - fQueenBmp.Height) div 2
  );
  QueenRect := Rect(0, 0, fQueenBmp.Width, fQueenBmp.Height);
  Canvas.Brush.Style := bsClear;
  for Square in fEngineState.Queens do
  begin
    TopLeft := Point(
      SquareSize * (Square.Col - 1),
      SquareSize * (Square.Row - 1)
    );
    Canvas.BrushCopy(
      Bounds(
        TopLeft.X + QueenOffset.X, TopLeft.Y + QueenOffset.Y,
        fQueenBmp.Width, fQueenBmp.Height
      ),
      fQueenBmp,
      QueenRect,
      clSilver  // makes silver background of queen bmp transparent
    );
  end;
end;

procedure TChessBoardCmp.SetBounds(ALeft, ATop, AWidth, AHeight: Integer);
begin
  // ignore AWidth and AHeight: size of control is determined by size of a
  // chessboard square and can't be changed any other way
  inherited SetBounds(ALeft, ATop, BoardSize, BoardSize);
end;

procedure TChessBoardCmp.SetSquareSize(const Value: TSquareSize);
begin
  if fSquareSize = Value then
    Exit;
  fSquareSize := Value;
  Invalidate;
end;

procedure TChessBoardCmp.ShowInfluences(const Square: TSquare);
begin
  fInfluencesSquare := Square;
  fShowInfluences := True;
  Invalidate;
end;

function TChessBoardCmp.SquareCoordsAt(const X, Y: Integer): TPoint;
begin
  Result.X := X div SquareSize + Low(TColCoord);
  Result.Y := Y div SquareSize + Low(TRowCoord);
end;

procedure TChessBoardCmp.UpdateBoard(const EngineState: TEngineState);
begin
  fEngineState := EngineState;
  Invalidate;
end;

end.



Source/FmMain.pas
{
 * Distributed under the MIT license.
 * See the accompanying LICENSE file or go to
 * http://delphidabbler.mit-license.org/1991-2016/
 *
 * $Rev: 66 $
 * $Date: 2016-02-14 02:15:25 +0000 (Sun, 14 Feb 2016) $
 *
 * 8 Queens application's main form.
}


unit FmMain;


interface


uses
  // Delphi
  ExtCtrls, ComCtrls, StdCtrls, Controls, Classes, Forms,
  // DelphiDabbler library
  PJVersionInfo, PJAbout, PJHotLabel,
  // Project
  UCalc, UChessBoardCmp;

type

  ///  <summary>Application's main form. Handles main window UI and interacts
  ///  with problem solving engine, displaying results.</summary>
  TQueensForm = class(TForm)
    btnAbout: TButton;
    btnAuto: TButton;
    btnExit: TButton;
    btnNext: TButton;
    btnReset: TButton;
    bvlBoard: TBevel;
    dlgAbout: TPJAboutBoxDlg;
    hlblWebsite: TPJHotLabel;
    lblAutoSpeed: TLabel;
    lblAutoSpeedSel: TLabel;
    lblSolutions: TLabel;
    rePrompt: TRichEdit;
    tbAuto: TTrackBar;
    tiAuto: TTimer;
    viMain: TPJVersionInfo;
    ///  <summary>Responds to About button click. Displays About box.</summary>
    procedure btnAboutClick(Sender: TObject);
    ///  <summary>Responds to Slideshow button click. Toggles slideshow on and
    ///  off.</summary>
    procedure btnAutoClick(Sender: TObject);
    ///  <summary>Responds to Exit button click. Closes program.</summary>
    procedure btnExitClick(Sender: TObject);
    ///  <summary>Responds to Next Solution button click. Displays next solution
    ///  or "finished" message.</summary>
    procedure btnNextClick(Sender: TObject);
    ///  <summary>Responds to Reset button click. Re-initialize program to
    ///  start-up state.</summary>
    procedure btnResetClick(Sender: TObject);
    ///  <summary>Initialises form's components and creates solution engine
    ///  object.</summary>
    procedure FormCreate(Sender: TObject);
    ///  <summary>Tears down solution engone object.</summary>
    procedure FormDestroy(Sender: TObject);
    ///  <summary>Updates slide show speed in reponse to change in associated
    ///  track bar.</summary>
    procedure tbAutoChange(Sender: TObject);
    ///  <summary>Handles slide show timer events by getting next solution from
    ///  solution engine.</summary>
    procedure tiAutoTimer(Sender: TObject);
  private
    ///  <summary>Component used to display chessboard and contents.</summary>
    fChessBoardCmp: TChessBoardCmp;
    ///  <summary>Problem solving engine.</summary>
    fEngine: TEngine;
    ///  <summary>Code giving current mode or state of program: waiting to start
    ///  (reset), manual mode, auto (slideshow) mode or finished (all solutions
    ///  displayed)</summary>
    fMode: Integer;
    ///  <summary>Gets next solution and displays it.</summary>
    procedure NextSolution;
    ///  <summary>Stops automatic generation of solutions ("slide show").
    ///  </summary>
    procedure AutoStop;
    ///  <summary>Starts automatic generation of solutions ("slide show").
    ///  </summary>
    procedure AutoStart;
    ///  <summary>Initialises display and solution engine.</summary>
    procedure Initialize;
    ///  <summary>Extracts rich text message from resource with given ID and
    ///  loads it into rich edit control.</summary>
    procedure LoadMessage(const ID: Integer);
    ///  <summary>Sets diaplay mode to given value and updates display and
    ///  solution engine accordingly.</summary>
    procedure SetMode(Mode: Integer);
    ///  <summary>Updates speed of "slide shows" and displays new speed.
    ///  </summary>
    procedure UpdateAutoCalcSpeed;
    ///  <summary>Updates chessboard to display current solution state.
    ///  </summary>
    procedure UpdateBoard;
    ///  <summary>Handles chessboard control's OnMouseDown event to display
    ///  influence lines of any square under mouse cursor.</summary>
    procedure BoardMouseDown(Sender: TObject; Button: TMouseButton;
      Shift: TShiftState; X, Y: Integer);
    ///  <summary>Handles chessboard control's OnMouseUp event to switch off
    ///  display of any influence lines.</summary>
    procedure BoardMouseUp(Sender: TObject; Button: TMouseButton;
      Shift: TShiftState; X, Y: Integer);
  end;

var
  QueensForm: TQueensForm;


implementation


uses
  // Delphi
  SysUtils, Windows,
  // Project
  UPlatform;

{$R *.DFM}

const
  // Programs modes (all used as IDs for help RTF messages displayed for each
  // mode)
  cModeReset    = 1;  // waiting to start finding solutions
  cModeManual   = 2;  // manual mode: user triggers finding solutions
  cModeAuto     = 3;  // auto mode: slide show triggers finding solutions
  cModeFinished = 4;  // finished: all solutions found

  // Map of slide show speed track bar's values onto description of speed and
  // time interval between display of solutions
  cAutoSpeeds: array[0..4] of record
    Desc: string;     // description of selected speed
    Delay: Integer;   // time interval for slide show timer
  end =
  (
    (Desc: 'Very Slow'; Delay: 2400;),
    (Desc: 'Slow';      Delay: 1800;),
    (Desc: 'Medium';    Delay: 1000;),
    (Desc: 'Fast';      Delay: 500;),
    (Desc: 'Very Fast'; Delay: 200;)
  );

resourcestring
  // Button captions
  sAutoStart = 'Start &Slide Show';
  sAutoStop = 'Stop &Slide Show';
  sManualStart = '&First Solution';
  sManualNext = '&Next Solution';
  // Solution label captions
  sSolutionStart = 'Ready';
  sSolutionStop = 'Finished';
  sSolutionCounter = 'Solution #%d';


{ TQueensForm }

procedure TQueensForm.AutoStart;
begin
  NextSolution;
  tiAuto.Enabled := True;
  btnAuto.Caption := sAutoStop;
end;

procedure TQueensForm.AutoStop;
begin
  tiAuto.Enabled := False;
  btnAuto.Caption := sAutoStart;
end;

procedure TQueensForm.BoardMouseDown(Sender: TObject; Button: TMouseButton;
  Shift: TShiftState; X, Y: Integer);
begin
  if not fChessBoardCmp.IsSquareAt(X, Y) then
    Exit;
  fChessBoardCmp.ShowInfluences(fChessBoardCmp.GetSquareAt(X, Y));
end;

procedure TQueensForm.BoardMouseUp(Sender: TObject; Button: TMouseButton;
  Shift: TShiftState; X, Y: Integer);
begin
  fChessBoardCmp.HideInfluences;
end;

procedure TQueensForm.btnAboutClick(Sender: TObject);
begin
  dlgAbout.Execute;
end;

procedure TQueensForm.btnAutoClick(Sender: TObject);
begin
  if fMode = cModeFinished then
    Exit;
  if fMode <> cModeAuto then
    SetMode(cModeAuto)    // start slideshow
  else
    SetMode(cModeManual); // stop slideshow
end;

procedure TQueensForm.btnExitClick(Sender: TObject);
begin
  Close;
end;

procedure TQueensForm.btnNextClick(Sender: TObject);
begin
  if fMode = cModeFinished then
    Exit;
  if fMode <> cModeAuto then
    SetMode(cModeManual);
  NextSolution;
end;

procedure TQueensForm.btnResetClick(Sender: TObject);
begin
  Initialize;
end;

procedure TQueensForm.FormCreate(Sender: TObject);
begin
  // Make use OSs default font
  SetDefaultFormFont(Self);
  // adjust hot label to use system font for highlighting
  hlblWebsite.HighlightFont.Name := hlblWebsite.Font.Name;
  hlblWebsite.HighlightFont.Size := hlblWebsite.Font.Size;
  // ensure hot label large enough and realign right
  hlblWebsite.AutoSize := True;
  hlblWebsite.Left := btnReset.Left + btnReset.Width - hlblWebsite.Width;

  // Window caption is application title
  Caption := Application.Title;

  // Create chessboard display
  fChessBoardCmp := TChessBoardCmp.Create(Self);
  fChessBoardCmp.Parent := Self;
  fChessBoardCmp.SquareSize := 24;
  fChessBoardCmp.Top := bvlBoard.Top + 1;
  fChessBoardCmp.Left := bvlBoard.Left + 1;
  fChessBoardCmp.OnMouseDown := BoardMouseDown;
  fChessBoardCmp.OnMouseUp := BoardMouseUp;

  // Create calculation engine
  fEngine := TEngine.Create;

  // Initialize engine and display
  Initialize;

  // Set default slide show speed
  UpdateAutoCalcSpeed;
end;

procedure TQueensForm.FormDestroy(Sender: TObject);
begin
  // Free objects
  fEngine.Free;
end;

procedure TQueensForm.Initialize;
begin
  SetMode(cModeReset);
end;

procedure TQueensForm.LoadMessage(const ID: Integer);
var
  Stm: TResourceStream; // stream onto required resource
begin
  // Help messages are stored in RCDATA resource with a numeric ID
  Stm := TResourceStream.CreateFromID(HInstance, ID, RT_RCDATA);
  try
    rePrompt.Lines.LoadFromStream(Stm);
  finally
    Stm.Free;
  end;
end;

procedure TQueensForm.NextSolution;
begin
  // Try to generate next solution
  if fEngine.NextSolution then
  begin
    // Solution found: update display
    lblSolutions.Caption := Format(sSolutionCounter, [fEngine.SolutionCount]);
    UpdateBoard;
  end
  else
    // No solution found: we have finished
    SetMode(cModeFinished);
end;

procedure TQueensForm.SetMode(Mode: Integer);
begin
  // Only proceed if mode has changed
  if fMode <> Mode then
  begin
    // Record new mode
    fMode := Mode;
    // Load help message with same ID as mode
    LoadMessage(Mode);
    // Update display according to mode
    case Mode of
      cModeReset:
      begin
        // Reset the program
        // halt any slide show
        AutoStop;
        // update buttons
        btnNext.Caption := sManualStart;
        btnNext.Enabled := True;
        btnAuto.Enabled := True;
        btnReset.Enabled := False;
        // update solution label to indicate ready to start
        lblSolutions.Caption := sSolutionStart;
        // reset solution engine and redraw (blank) chessboard
        fEngine.Initialize;
        UpdateBoard;
      end;
      cModeManual:
      begin
        // Change to manual mode
        // halt any slide show
        AutoStop;
        // update buttons
        btnNext.Caption := sManualNext;
        btnReset.Enabled := True;
      end;
      cModeAuto:
      begin
        // Change to auto (slide show) mode
        // update buttons
        btnNext.Caption := sManualNext;
        btnReset.Enabled := True;
        // start the slide show
        AutoStart;
      end;
      cModeFinished:
      begin
        // Change to finished mode (all solutions found)
        // halt any slide show
        AutoStop;
        // update buttons
        btnNext.Enabled := False;
        btnAuto.Enabled := False;
        btnReset.Enabled := True;
        // update solution label to indicate finished
        lblSolutions.Caption := sSolutionStop;
        // redraw (blank) chess board
        UpdateBoard;
      end;
    end;
  end;
end;

procedure TQueensForm.tbAutoChange(Sender: TObject);
begin
  UpdateAutoCalcSpeed;
end;

procedure TQueensForm.tiAutoTimer(Sender: TObject);
begin
  NextSolution; // this method will switch modes if no more solutions
end;

procedure TQueensForm.UpdateAutoCalcSpeed;
begin
  lblAutoSpeed.Caption := cAutoSpeeds[tbAuto.Position].Desc;
  tiAuto.Interval := cAutoSpeeds[tbAuto.Position].Delay;
end;

procedure TQueensForm.UpdateBoard;
begin
  fChessBoardCmp.UpdateBoard(fEngine.EngineState);
end;

end.

